lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -1331,21 +1331,29 @@ void DifferentiableActivityInfo::analyze(DominanceInfo *di,
           if (isVaried(cai->getSrc(), i))
             recursivelySetVaried(cai->getDest(), i);
         }
-        // Handle `struct_extract`.
-        else if (auto *sei = dyn_cast<StructExtractInst>(&inst)) {
-          if (isVaried(sei->getOperand(), i)) {
-            // If `@noDerivative` exists on the field while the struct is
-            // `@_fieldwiseDifferentiable`, this field is not in the set of
-            // differentiable variables that we want to track the variedness of.
-            auto hasNoDeriv = sei->getField()->getAttrs()
-                .hasAttribute<NoDerivativeAttr>();
-            auto structIsFieldwiseDiffable = sei->getStructDecl()->getAttrs()
-                .hasAttribute<FieldwiseDifferentiableAttr>();
-            if (!(hasNoDeriv && structIsFieldwiseDiffable))
-              for (auto result : inst.getResults())
-                setVaried(result, i);
-          }
-        }
+
+// Handle `struct_extract` and `struct_element_addr` instructions.
+// - If the field is marked `@noDerivative` and belongs to a
+//   `@_fieldwiseDifferentiable` struct, do not set the result as varied because
+//   it is not in the set of differentiable variables.
+// - Otherwise, propagate variedness from operand to result as usual.
+#define VISIT_STRUCT_ELEMENT_INST(INST) \
```

Comment:
```suggestion
#define PROPAGATE_VARIED_FOR_STRUCT_EXTRACTION(INST) \
```

Reply:
Done in https://github.com/apple/swift/pull/22973/commits/e42706e5b266d61589a06dcb3ee67cfc1a2ac407.